### PR TITLE
Fix `__errno_location` under wasm workers. NFC

### PR DESCRIPTION
**Pull request by sbc100**
Previously the wasm workers build of `__errno_location` was using the
single threaded code path which doesn't use a thead local location.

In addition this change avoids the use of `__EMSCRIPTEN_PTHREADS__` in
`__errno_location.c` which is helpful for unifying the shared memory
libc build.  See #22735.

### CODE REVIEW

**Comment by sbc100** [comment]
@eyebrowsoffire

**Comment by sbc100** [comment]
@eyebrowsoffire I'm running into the fact the `-sSHARED_MEMORY` builds don't think here..  :(   I guess we need your change first?

File: system/lib/libc/musl/src/errno/__errno_location.c [hunk 1]
```diff
@@ -1,22 +1,19 @@
...
-{
-	return &__pthread_self()->errno_val;
-}
-#else
-// for single-threaded mode, avoid linking in pthreads support code
-// just for this
-static int __errno_storage = 0;
+#if __EMSCRIPTEN__
+// For emscripten we use TLS here instead of `__pthread_self`, so that in single
+// threaded builds this get lowered away to normal global variable.
```
Comment (kripken) [inline system/lib/libc/musl/src/errno/__errno_location.c:6]:
```suggestion
// threaded builds this gets lowered away to a normal global variable.
```

File: test/test_sockets.py [hunk 2]
```diff
@@ -328,7 +328,7 @@ def test_nodejs_sockets_echo_subprotocol(self):
   # NOTE: Shared buffer is not allowed for websocket sending.
   @parameterized({
     '': [[]],
-    'shared': [['-sSHARED_MEMORY']],
+    'pthread': [['-pthread']],
```
Comment (kripken) [inline test/test_sockets.py:331]: Is this change related to errno?

**Review by kripken — APPROVED** [review]

File: test/code_size/hello_wasm_worker_wasm.json [hunk 3]
```diff
@@ -5,8 +5,8 @@
   "a.js.gz": 455,
   "a.ww.js": 115,
   "a.ww.js.gz": 127,
-  "a.wasm": 1850,
-  "a.wasm.gz": 1050,
-  "total": 3248,
-  "total_gz": 2016
+  "a.wasm": 1894,
```
Comment (kripken) [inline test/code_size/hello_wasm_worker_wasm.json:8]: Why did this grow by 44 bytes?

Comment (sbc100) [reply test/code_size/hello_wasm_worker_wasm.json:8]: I think because it fixes the wasm workers implementation of errno_location so that it returns a TLS location, which I think is bug fix.

Comment (sbc100) [reply test/code_size/hello_wasm_worker_wasm.json:8]: Prior to this change the WW version of errno_location was the same as the non-threaded version (i.e. not thread local or thread safe).

**Comment by sbc100** [comment]
Updated PR description and title to better reflect the purpose of this change.

File: test/code_size/hello_wasm_worker_wasm.json (hunk 3 above)
Comment (kripken) [reply test/code_size/hello_wasm_worker_wasm.json:8]: Ah, right... good fix then.

File: tools/system_libs.py [hunk 4]
```diff
@@ -742,7 +742,7 @@ def vary_on(cls):
 
   @classmethod
   def get_default_variation(cls, **kwargs):
-    return super().get_default_variation(is_mt=settings.PTHREADS, is_ww=settings.WASM_WORKERS and not settings.PTHREADS, **kwargs)
+    return super().get_default_variation(is_mt=settings.PTHREADS, is_ww=settings.SHARED_MEMORY and not settings.PTHREADS, **kwargs)
```
Comment (kripken) [inline tools/system_libs.py:745]: Is this needed in this PR?

**Review by kripken — APPROVED** [review]
lgtm otherwise

File: tools/system_libs.py (hunk 4 above)
Comment (sbc100) [reply tools/system_libs.py:745]: @eyebrowsonfire, is this maybe enough to resolve https://github.com/emscripten-core/emscripten/pull/22683

Comment (sbc100) [reply tools/system_libs.py:745]: @eyebrowsoffire, is this maybe enough to resolve #22683

Comment (sbc100) [reply tools/system_libs.py:745]: The reason this is needed for this PR is because it makes `__errno_location` into a function that, in non-threads build, has TLS erased, making it incompatible with `-sSHARED_MEMORY`, which triggers the problem found in #22683.   Basically it make the problem found in #22683 happen is way more cases, including simple hello world programs.
